submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,250 @@
+|:---------------------------|:-------------------------------------------------------------------------------|
+| `--kubeconfig` `<string>`  | Absolute path(s) to the kubeconfig file(s) (default `$HOME/.kube/config`)
+| `--kubecontext` `<string>` | Kubeconfig context to use
+| `--name` `<string>`        | The name of the policy to delete
+
+##### `list`
+
+`subctl cable-policy list [flags]`
+
+Inspects the cluster and reports information about the detected cable policies.
+
+| Flag                         | Description
+|:-----------------------------|:----------------------------------------------------------------------------|
+| `--kubeconfig` `<string>`    | Absolute path(s) to the kubeconfig file(s) (default `$HOME/.kube/config`)
+| `--kubecontext` `<string>`   | Kubeconfig context to use
+
+#### Connection Policy Examples
+
+- we want to specify a default policy that uses IPsec as the default cable driver.
+
+```yaml
+kind: CablePolicies
+metadata:
+  name: default
+spec:
+  cableDriver:
+    name: "ipsec"
+```
+
+- We want to use specific cable drivers to connect clusters in a ClusterSet as shown in the image below. We
+label the `Clusters` appropriately and define a number of policies with the appropriate label selectors.
+
+![Finer grained connectivity example](./images/multiple_cables_ep.png)
+
+```yaml
+kind: CablePolicies
+metadata:
+  name: default
+spec:
+  cableDriver:
+    name:  "vxlan"
+```
+
+```yaml
+kind: CablePolicies
+metadata:
+  name: on-prem-to-cloud
+spec:
+  leftClusterSelector:
+    env: production
+  rightClusterSelector:
+    env: production
+  cableDriver:
+    name:  "ipsec"
+```
+
+### Cable driver configuration
+
+The cable driver configuration in Submariner only allows for the configuration of a single cable driver for a
+gateway. This will pose a challenge when allowing for varying (or multiple) cable types between clusters based
+on the defined cable driver policies. The goals of this part of the proposal is to extend the cable driver
+configuration to:
+
+- Maintain backward compatibility with the current cable driver configuration
+- Create an extended cable driver configuration that allows for the setup of one, many or combined drivers.
+
+The proposed approach for additional Submariner cable driver configuration parameters is:
+
+- To continue to use the existing environment variables injected into the Gateway Pod (with the exception of PSKs/secrets).
+  These environment variables will come from the `Submariner.io_Submariners.yaml`.
+- To extend `Submariner.io_Submariners.yaml` with some new global parameters for multiple cable driver configurations
+  in a ClusterSet.
+- To use a volume mounted ConfigMap for more detailed cable driver configurations (as these may vary and should be shared across
+  gateway instances in the ClusterSet).
+- To use mounted secrets for parameters that need to be protected, e.g. the Pre-Shared Key (PSK) and to abandon the use of
+  environment variables for secrets.
```

Comment:
All the info needed to connect clusters should be in the CRDs so I was imagining the tunnel controller in the GW would set up cable drivers based on that. It could read all the policies on startup and start the referenced drivers prior to processing `Endpoints` but it's going to have to react to policy changes as well so it's going to have to be able to spin up cable driver instances on the fly anyway so I was thinking we may as well handle startup in the same manner. So, on startup, no cable drivers are started. When an `Endpoint` is received, it consults the policies to determine which driver to use and spins it up if not already started.